### PR TITLE
CORE-16230 Split UniquenessProcessor into its own worker

### PR DESCRIPTION
**Pull request by mbrkic-r3**
The purpose of this task is to build a new UniquenessWorker around UniquenessProcessor, removing it from the DBWorker (see CORE-16230 for more info).

Related PRs:

- https://github.com/corda/corda-shared-build-pipeline-steps/pull/1061
- https://github.com/corda/corda-e2e-tests/pull/194
- https://github.com/corda/corda-non-functional-test/pull/224
- https://github.com/corda/corda-soak-test-deployment/pull/122

Resiliency tests:
- https://github.com/corda/corda-soak-test-deployment/compare/mbrkic-r3/CORE-16230/uniq?expand=1
- https://ci02.dev.r3.com/blue/organizations/jenkins/Corda5%2Fcorda-resiliency-tests/detail/mbrkic-r3%2FCORE-16230%2Funiq/4/pipeline/

Merging strategy:

It was agreed with BLT that this is merged using feature branch of `corda-shared-build-pipeline-steps`, after that the shared pipeline will be merged (deleting feature branch is temporarily disabled) and then `corda-runtime-os` will be modified to use 5.1 branch of shared pipeline and that PR merged as well.

### CODE REVIEW

**Comment by chrisr3** [comment]
Why split the uniqueness processor into its own worker?

**Comment by mbrkic-r3** [comment]
Flows will communicate with other workers via RPC, not Kafka. Having more different processors in different workers will enable scaling each of them as separately. I'm not sure whether/how much difference will it actually make, @Omar-awad did a POC and can comment more regarding this.

**Comment by chrisr3** [comment]
> Flows will communicate with other workers via RPC, not Kafka. Having more different processors in different workers will enable scaling each of them as separately. I'm not sure whether/how much difference will it actually make, @Omar-awad did a POC and can comment more regarding this.

At the moment, each flow processor is deployed alongside its own uniqueness processor within the same JVM, and I really don't see how you can get any faster than that.

**Comment by relyafi** [comment]
> Flows will communicate with other workers via RPC, not Kafka. Having more different processors in different workers will enable scaling each of them as separately. I'm not sure whether/how much difference will it actually make, @Omar-awad did a POC and can comment more regarding this.

@mbrkic-r3 that doesn't sound like a strong reason to me - I thought we'd established during the GA release cycle about the need to show a benefit from performance related changes? Can you point me to the evidence of the performance testing you've done to show that this change will be a positive one?

From conversations I've had with @driessamyn and @Omar-awad it sounds like there are other reasons for making this change, but the difficulty is this hasn't been properly communicated with the SME's, and I need that background in order to sensibly review this PR. Can we get a meeting set up with yourself / @JamesHR3 / @Omar-awad to discuss the background in more detail?

> At the moment, each flow processor is deployed alongside its own uniqueness processor within the same JVM, and I really don't see how you can get any faster than that.

@chrisr3 not quite - it actually lives as part of the DB worker alongside the DB processor, with the idea being that it is then co-located with DB access functions which matter given how DB sensitive the uniqueness checker is. But I agree with the thrust of your argument.

**Comment by chrisr3** [comment]
> @chrisr3 not quite - it actually lives as part of the DB worker alongside the DB processor, with the idea being that it is then co-located with DB access functions which matter given how DB sensitive the uniqueness checker is. But I agree with the thrust of your argument.

OK. The problem with writing something like the driver is that you end up finding and using all the necessary components, but without always understanding where they fit into a _real_ Corda deployment :wink:.

But regardless, I had thought we'd discovered that our Kafka problem was _at least in part_ due to our unnecessarily sending messages over Kafka that could have been handled locally.

As an aside, I have previously (i.e. before R3) written servers for a micro-service architecture that was designed to communicate via HTTP(S), and IIRC it was very much an "eventually consistent" system requiring idempotent messaging. And it wasn't trivial to write either!

**Comment by Omar-awad** [comment]
>I thought we'd established during the GA release cycle about the need to show a benefit from performance related changes? Can you point me to the evidence of the performance testing you've done to show that this change will be a positive one?

This change will not have performance improvements by itself as it is part of a bigger initiative which is documented in high level Design. As of why we need this segregation then please refer to another PR comment here: https://github.com/corda/corda-runtime-os/pull/4465#issuecomment-1683769408 

The two main issues we are tackling in the next release are:

 - Optimize the events flow in the system by reducing loops and congestion on some topics.
 - Revise heavy communication protocols in an effort to reduce the number of messages over-the-wire.

**Comment by chrisr3** [comment]
> > I thought we'd established during the GA release cycle about the need to show a benefit from performance related changes? Can you point me to the evidence of the performance testing you've done to show that this change will be a positive one?
> 
> This change will not have performance improvements by itself as it is part of a bigger initiative which is documented in high level here: https://r3-cev.atlassian.net/wiki/spaces/CB/pages/4552491009/Corda+Scalable+Topology+Design. As of why we need this segregation then please refer to another PR comment here: [#4465 (comment)](https://github.com/corda/corda-runtime-os/pull/4465#issuecomment-1683769408)

Actually, I think @relyafi was asking to examine the performance data that presumably underlies all these extensive architectural changes. I don't think that particular comment qualifies.

**Comment by Omar-awad** [comment]
>Actually, I think [@relyafi](https://github.com/relyafi) was asking to examine the performance data that presumably underlies all these extensive architectural changes. I don't think that particular comment qualifies.

Please refer to the design document. The reasoning and the verification is documented there.

**Comment by relyafi** [comment]
@Omar-awad:

> This change will not have performance improvements by itself as it is part of a bigger initiative which is documented in high level here: https://r3-cev.atlassian.net/wiki/spaces/CB/pages/4552491009/Corda+Scalable+Topology+Design.

Thanks for forwarding that document. The overall theoretical approach seems reasonable, but I don't see it mention splitting the uniqueness worker as one of its objectives? Combined with the fact that you say this has no performance benefits in and of itself, why do we feel this change is necessary at this point?

> The two main issues we are tackling in the next release are:
> 
> * Optimize the events flow in the system by reducing loops and congestion on some topics.
> * Revise heavy communication protocols in an effort to reduce the number of messages over-the-wire.

I don't see how this change addresses either of these issues? Uniqueness already has its own topic, and this change doesn't do anything with communication protocols.

> Please refer to the design document. The reasoning and the verification is documented there.

The only evidence I see in that document is a PoC, which I'm assuming has made a multitude of changes? It could be that the individual changes made in the PoC were a mix of positive, neutral, and negative, and on balance they turned out to be positive. That does not mean we can infer that every change in the PoC was a positive.

The only way to do that is A/B testing for a specific change we want to make. I think the following from the acceptance criteria of the document you linked backs this up:

_Implemented components should show clear value in being shipped._

As it stands, I don't think this PR passes that test.

**Review by driessamyn — APPROVED** [review]
Change LGTM and lines up with the intended design.

**Review by relyafi — DISMISSED** [review]
See Jira ticket for my comments.

**Review by relyafi — COMMENTED** [review]
See Jira ticket for my comments.

**Review by relyafi — COMMENTED** [review]
See Jira ticket for my comments.

**Review by Omar-awad — APPROVED** [review]
All good.

**Review by ben-millar — APPROVED** [review]
LGTM

File: .ci/JenkinsfileSnykDelta [hunk 1]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/JenkinsfileSnykDelta:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/PublishHelmChart/Jenkinsfile_PublishHelmChart [hunk 2]
```diff
@@ -1,5 +1,7 @@
 #! groovy
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/PublishHelmChart/Jenkinsfile_PublishHelmChart:2]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/e2eTests/Jenkinsfile [hunk 3]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/e2eTests/Jenkinsfile:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/e2eTests/JenkinsfileCombinedWorker [hunk 4]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/e2eTests/JenkinsfileCombinedWorker:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/e2eTests/JenkinsfileCombinedWorkerLinux [hunk 5]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/e2eTests/JenkinsfileCombinedWorkerLinux:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/e2eTests/JenkinsfileCombinedWorkerWindows [hunk 6]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/e2eTests/JenkinsfileCombinedWorkerWindows:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/e2eTests/JenkinsfileMultiClusterTest [hunk 7]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/e2eTests/JenkinsfileMultiClusterTest:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/e2eTests/JenkinsfileUnstableTest [hunk 8]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/e2eTests/JenkinsfileUnstableTest:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/e2eTests/ethereum/Jenkinsfile [hunk 9]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/e2eTests/ethereum/Jenkinsfile:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/nightly/JenkinsfileNightly [hunk 10]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/nightly/JenkinsfileNightly:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/nightly/JenkinsfileSnykScan [hunk 11]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/nightly/JenkinsfileSnykScan:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/nightly/JenkinsfileUnstableTests [hunk 12]
```diff
@@ -1,5 +1,7 @@
 //catch all job for flaky tests
-@Library('corda-shared-build-pipeline-steps@5.0.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/nightly/JenkinsfileUnstableTests:2]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/nightly/JenkinsfileWindowsCompatibility [hunk 13]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/nightly/JenkinsfileWindowsCompatibility:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/versionCompatibility/Jenkinsfile [hunk 14]
```diff
@@ -1,3 +1,5 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/versionCompatibility/Jenkinsfile:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: .ci/versionCompatibility/latest-version.Jenkinsfile [hunk 15]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline .ci/versionCompatibility/latest-version.Jenkinsfile:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

File: Jenkinsfile [hunk 16]
```diff
@@ -1,4 +1,6 @@
-@Library('corda-shared-build-pipeline-steps@5.1') _
+@Library('corda-shared-build-pipeline-steps@mbrkic-r3/CORE-16230/uniqueness-worker') _
```
Comment (mcgovc) [inline Jenkinsfile:1]:
```suggestion
@Library('corda-shared-build-pipeline-steps@5.1') _
```

**Review by mcgovc — APPROVED** [review]
LGTM

File: Jenkinsfile (hunk 16 above)
Comment (ronanbrowne) [reply Jenkinsfile:1]: Discussed here https://github.com/corda/corda-shared-build-pipeline-steps/pull/1061#issuecomment-1685923026 - will leave as is and address in second PR to minimize impacts

**Review by thiagoviana — APPROVED** [review]

**Review by josephzunigadaly — APPROVED** [review]

**Review by oscardalmau-r3 — APPROVED** [review]

**Review by blsemo — APPROVED** [review]
Looks ok to me.

**Pull request merged**